### DISCUSSION TITLE
Cannot perform single test in Intellij

### DISCUSSION THREAD

**Discussion by mosesonline** [mapstruct/mapstruct#3461] [Q&A]
Hello,

I would like to contribute, so I checked out the code and I can `mvn install` and use the build version. I can run all the tests with mvn but to understand the code I would like to debug it. 
So I tried to run a single test (AccessibilityTest#testGeneratedModifiersFromAbstractClassAreCorrect) but i recieve an error:
```
mapstruct\processor\src\main\java\org\mapstruct\ap\internal\util\AbstractElementUtilsDecorator.java:22:32
java: cannot find symbol
  symbol:   class ModuleElement
  location: package javax.lang.model.element
```
I use java corretto 11.0.21 for the mvn builds and the singel test runs.

**Comment by thunderhook** [comment]
I have the same problem, and we're not alone. I had an ugly solution which consisted of commenting out uncompilable code (see https://github.com/mapstruct/mapstruct/pull/3411#discussion_r1367720384). 
This thread encouraged me to take another look.

I found a setting in IntelliJ that would allow the code to compile:
* Navigate to [`File | Settings | Build, Execution, Deployment | Compiler | Java Compiler`](jetbrains://Idea/settings?name=Build%2C+Execution%2C+Deployment--Compiler--Java+Compiler)
* Set the target bytecode version of the `mapstruct-processor` module to at least 9.

![settings](https://github.com/mapstruct/mapstruct/assets/8238759/c22fa0ed-d31f-48ab-a333-27794803ed4f)

This setting will be undone when you reload the project from Maven. You will then need to do this setting again.
There is a [setting that prevents IntelliJ from using the target level from maven](https://intellij-support.jetbrains.com/hc/en-us/community/posts/115000347524/comments/360000030024), but this is a global flag and therefore not recommended as a potential fix.

Tell me if this has worked for you.

**Reply by mosesonline** [reply]
Thank you for the fast reply.

It sounded reasonable to me, but unfortunatly it's not working. Tried to re-index and turn it off and on but:

![grafik](https://github.com/mapstruct/mapstruct/assets/1179955/72c514a8-1b3f-46c2-b276-640909964580)

After i reconfigured it in maven to use 9 as target. The test starts but fails:
![grafik](https://github.com/mapstruct/mapstruct/assets/1179955/a5b7a5b4-60e3-4cc2-9002-ab5ee7a3ec00)

**Reply by filiphr** [reply]
Which SDK are you both using for the project?

![image](https://github.com/mapstruct/mapstruct/assets/6012094/955b94b6-f56c-4133-ab63-b3ff5e3cfa5e)


I am using JDK 21, you need to use at least JDK 17 for the code to compile in IntelliJ properly. I don't need to change the module language level that is Language Level 8 for me

![image](https://github.com/mapstruct/mapstruct/assets/6012094/0a91daab-e75e-4e9a-9d5c-bb21c90da857)

**Reply by thunderhook** [reply]
Using OpenJDK 17, language level is also 8 in my settings.

@filiphr Could you please check what is shown for you in the settings page `File | Settings | Build, Execution, Deployment | Compiler | Java Compiler`? This is the only way I can get it to work.

I have also tried to upgrading to OpenJDK 21. After a `mvn clean` (to make sure that there is no old compiled code left) I get strange errors (interestingly only in 60 tests), e.g:

```
SpringSetterMapperTest
org.springframework.beans.factory.BeanDefinitionStoreException: Failed to read candidate component class: file 
[C:\...\mapstruct\processor\target\compilation-tests\org.mapstruct.ap.test.injectionstrategy.spring.setter.SpringSetterMapperTest\shouldHaveSetterInjection_jdk\classes\org\mapstruct\ap\test\injectionstrategy\spring\setter\CustomerRecordSpringSetterMapper.class]; 
nested exception is org.springframework.core.NestedIOException: ASM ClassReader failed to parse class file - probably due to a new Java class file version that isn't supported yet: 
file [C:\...\mapstruct\processor\target\compilation-tests\org.mapstruct.ap.test.injectionstrategy.spring.setter.SpringSetterMapperTest\shouldHaveSetterInjection_jdk\classes\org\mapstruct\ap\test\injectionstrategy\spring\setter\CustomerRecordSpringSetterMapper.class]; 
nested exception is java.lang.IllegalArgumentException: Unsupported class file major version 65
```

Side note, interestingly `JodaConversionTest` also fails, is this also the case for you @filiphr?
```
org.opentest4j.AssertionFailedError: 
expected: "1. Januar 2014 um 00:00:00 UTC"
 but was: "1. Januar 2014, 00:00:00 UTC"
```
This seems like my compile phase does not pick up Java 21.

----

With OpenJDK 21 I also tried to delegate my tests to maven: 
![grafik](https://github.com/mapstruct/mapstruct/assets/8238759/7129b7a2-0f8c-4f39-a882-34bc49aaeb20)

However the same test failures occur. No matter if I use the bundled maven or the maven wrapper.
@filiphr How is your maven configured? Do you have a environment-wide configured maven (like, does it work in a terminal outside of IntelliJ)? For me, this is not the case on this machine.

**Reply by filiphr** [reply]
>@filiphr Could you please check what is shown for you in the settings page `File | Settings | Build, Execution, Deployment | Compiler | Java Compiler`? This is the only way I can get it to work.

I have here "Same as language level" and language level is 1.8.

> I get strange errors (interestingly only in 60 tests), e.g:

I believe that this is related to the fact that the code we generate during our tests is using the language level of the compiler (we don't see that I believe) and since we are using Spring 5 it cannot use those compiled classes. I'll try to fix that, I know what the issue might be. If I use JDK 17 as the SDK the test passes, if I use JDK 21 it fails.

>Side note, interestingly `JodaConversionTest` also fails, is this also the case for you @filiphr?

Interesting, with Java 17 the test passes, with Java 21 it fails. I think that they have changed the behaviour of the `SImpleDateFormatter` in Java 21. I'll fix that test.

----

Seems like Java 21 is the issue. I've created https://github.com/mapstruct/mapstruct/issues/3473 so that I can fix it

**Comment by zyberzebra** [comment]
Just saying: I still had to change the target bytecode Version:
![image](https://github.com/user-attachments/assets/e09ab3a2-f081-4f09-8cec-4b1ee3e2f2cd)
![image](https://github.com/user-attachments/assets/213f389f-b1d0-482f-ae4a-a2e383b84838)

**Reply by filiphr** [reply]
I really do not understand this. I've never had any issues and I have not messed with the settings for this. Are you using Community or Ultimate edition?

**Reply by zyberzebra** [reply]
Ultimate. Jeah I also don't know. I had the same problem on Windows and Mac
